// @redocly/cli/lib/commands/stats.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleStats = void 0;
const perf_hooks_1 = require("perf_hooks");
const colors = require("colorette");
const openapi_core_1 = require("@redocly/openapi-core");
const utils_1 = require("../utils");
const utils_2 = require("../utils");
const statsAccumulator = {
    refs: { metric: '🚗 References', total: 0, color: 'red', items: new Set() },
    externalDocs: { metric: '📦 External Documents', total: 0, color: 'magenta' },
    schemas: { metric: '📈 Schemas', total: 0, color: 'white' },
    parameters: { metric: '👉 Parameters', total: 0, color: 'yellow', items: new Set() },
    links: { metric: '🔗 Links', total: 0, color: 'cyan', items: new Set() },
    pathItems: { metric: '➡️ Path Items', total: 0, color: 'green' },
    operations: { metric: '👷 Operations', total: 0, color: 'yellow' },
    tags: { metric: '🔖 Tags', total: 0, color: 'white', items: new Set() },
};
function printStatsStylish(statsAccumulator) {
    for (const node in statsAccumulator) {
        const { metric, total, color } = statsAccumulator[node];
        process.stderr.write(colors[color](`${metric}: ${total} \n`));
    }
}
function printStatsJson(statsAccumulator) {
    const json = {};
    for (const key of Object.keys(statsAccumulator)) {
        json[key] = {
            metric: statsAccumulator[key].metric,
            total: statsAccumulator[key].total,
        };
    }
    process.stdout.write(JSON.stringify(json, null, 2));
}
function printStats(statsAccumulator, api, format) {
    process.stderr.write(`Document: ${colors.magenta(api)} stats:\n\n`);
    switch (format) {
        case 'stylish':
            printStatsStylish(statsAccumulator);
            break;
        case 'json':
            printStatsJson(statsAccumulator);
            break;
    }
}
function handleStats(argv) {
    return __awaiter(this, void 0, void 0, function* () {
        const config = yield utils_1.loadConfigAndHandleErrors({ configPath: argv.config });
        const [{ path }] = yield utils_1.getFallbackApisOrExit(argv.api ? [argv.api] : [], config);
        const externalRefResolver = new openapi_core_1.BaseResolver(config.resolve);
        const { bundle: document } = yield openapi_core_1.bundle({ config, ref: path });
        const lintConfig = config.styleguide;
        const oasVersion = openapi_core_1.detectOpenAPI(document.parsed);
        const oasMajorVersion = openapi_core_1.openAPIMajor(oasVersion);
        const types = openapi_core_1.normalizeTypes(lintConfig.extendTypes(oasMajorVersion === openapi_core_1.OasMajorVersion.Version3 ? openapi_core_1.Oas3Types : openapi_core_1.Oas2Types, oasVersion), lintConfig);
        const startedAt = perf_hooks_1.performance.now();
        const ctx = {
            problems: [],
            oasVersion: oasVersion,
            visitorsData: {},
        };
        const resolvedRefMap = yield openapi_core_1.resolveDocument({
            rootDocument: document,
            rootType: types.Root,
            externalRefResolver,
        });
        const statsVisitor = openapi_core_1.normalizeVisitors([
            {
                severity: 'warn',
                ruleId: 'stats',
                visitor: openapi_core_1.Stats(statsAccumulator),
            },
        ], types);
        openapi_core_1.walkDocument({
            document,
            rootType: types.Root,
            normalizedVisitors: statsVisitor,
            resolvedRefMap,
            ctx,
        });
        printStats(statsAccumulator, path, argv.format);
        utils_2.printExecutionTime('stats', startedAt, path);
    });
}
exports.handleStats = handleStats;
